lib/roar/json/json_api.rb

```diff
@@ -152,17 +153,75 @@ def to_document(res, options)
         end
 
         def from_document(hash)
-          hash["data"]
+          if hash["data"].is_a?(Array)
+            hash["data"].map(&method(:from_document_data))
+          else
+            from_document_data hash["data"]
+          end
+        end
+
+        def from_document_data(data)
+          result = data.merge data.delete("attributes").to_h
+
+          if rels = result.delete("relationships")
+            result["links"] ||= {}
+
+            rels.each { |k, v|
+              if v["data"].is_a?(Array)
+                result["links"][k] = v["data"].map { |e| e["id"] }
+              else
+                result["links"][k] = v["data"]["id"]
+              end
+            }
+          end
+
+          result
+        end
+
+        def prepare_res(res)
+          extracted = extract_identifiers(res)
+
+          relationships = prepare_rels(res.delete "included") if res.key? "included"
+          extracted["attributes"] = res unless res.empty?
+          extracted["relationships"] = relationships if relationships
+
+          extracted
+        end
+
+        def prepare_rels(rels)
+          final = {}
+          rels.each { |k, v|
+            if v.is_a?(Array)
```

Comment:
Why are there so many `is_a?(Array)` here? Is JSON-API that fuzzy about its syntax? Isn't there a way to tell whether a hash entity is singular or array from the document specification (e.g. `property` vs. `collection`)?